scarlet/wavelet.py

```diff
@@ -0,0 +1,300 @@
+import autograd.numpy as np
+from . import fft
+from . import Cache
+from scipy.stats import median_absolute_deviation as mad
+
+# Filter for the scarlet transform. Here bspline
+h = np.array([1. / 16, 1. / 4, 3. / 8, 1. / 4, 1. / 16])
+
+class Starlet(object):
+    """ A class used to create the Wavelet transform of a cube of images from the 'a trou' algorithm.
+
+        The transform is performed by convolving the image by a seed starlet: the transform of an all-zero
+        image with its central pixel set to one. This requires 2-fold padding of the image and an odd pad
+        shape. The fft of the seed starlet is cached so that it can be reused in the transform of other
+        images that have the same shape.
+    """
+    def __init__(self, image, lvl = None, starlet = None, direct = False):
+        """ Initialise the Starlet object
+
+        Paramters
+        image: Fourier object
+            image to transform
+        lvl: int
+            number of starlet levels to use in the decomposition
+        starlet: array
+            Starlet transform of an array
+        direct: bool
+            if set to True, uses direct wavelet transform with the a trou algorithm.
+            if set to False, the transform is performed by convolving the image by the wavelet transform of a dirac.
+        """
+        # Shape for the starlet padding. It is also an fft fast shape.
+        self.seed = None
+        if starlet is None:
+            self._starlet_shape = get_starlet_shape(image.shape, lvl = lvl)
+        else:
+            self._starlet_shape = starlet.shape
+        # Image (as a cube)
+        if len(image.shape) == 2:
+            self._image = image[np.newaxis, :, :]
+        else:
+            self._image = image
+        # Original shape of the image
+        self._image_shape = self._image.shape
+        # Can be initialised by a starlet transform to perform inverse transforms
+        if starlet is None:
+            if direct == True:
+                self._starlet = self.direct_transform()
+            else:
+                self._starlet = self.transform()
+        else:
+            if len(np.shape(starlet)) == 3:
+                self._starlet = starlet[np.newaxis, :, :, :]
+            else:
+                self._starlet = starlet
+
+        if self.seed is None:
+            self.seed = mk_starlet(self._starlet_shape)
+        self._norm = np.sqrt(np.sum(self.seed ** 2, axis=(-2, -1)))
+
+    @property
+    def image(self):
+        """The real space image"""
+        return self._image
+
+    @property
+    def norm(self):
+        """The norm of the seed wavelet in each wavelet level (not in coarse wavelet)"""
+        return self._norm
+
+
+    @property
+    def starlet(self):
+        """The real space image"""
```

Comment:
I doubt that this docstring is correct

Reply:
Also, I dislike that the method name is a lower-case version of the class name. This interferes with the convention of using lower-case names for instances of that class, e.g. `starlet = Starlet()`. Can you pick a suitable name that also clarifies what the method will return.

Reply:
`coefficients` it is